### PR TITLE
Add support for kerberized Thrift Servers

### PR DESCRIPTION
**Pull request by QwertyManiac**
Since Apache Thrift 0.9.2 onwards includes a Python module for SASL,
we can use the TSaslClientTransport class to make secure connections.

This requires manual installation of "puresasl" and "kerberos" modules
to fully work. Since the "kerberos" module will only work in Python 2.7+,
this feature will not work for users of Python 2.6 or below.

I tested the change with the simple HBase Thrift Server setup of `auth` and also `auth-conf`, and both of these modes work with the changes made.

### CODE REVIEW

**Comment by QwertyManiac** [comment]
This wouldn't work until Thriftpy adds the TSaslClientTransport. I didn't notice when porting my 0.9 patch that we've shifted to use it.

I wonder why we use that module when the direct Thrift's Python library is also pure Python?

**Comment by QwertyManiac** [comment]
For 0.9 users, they can do this to immediately get things working (happybase example is reused in the code below, but ensure to replace HOSTNAME_OF_THRIFT_SERVER with your hostname, and to create the HBase table `t1` with family `family` before running this):

```
~> cat > foo.py
import happybase

connection = happybase.Connection('HOSTNAME_OF_THRIFT_SERVER', use_kerberos=True)
table = connection.table('t1')

table.put('row-key', {'family:qual1': 'value1',
                      'family:qual2': 'value2'})

row = table.row('row-key')
print row['family:qual1']  # prints 'value1'

for key, data in table.rows(['row-key-1', 'row-key-2']):
    print key, data  # prints row key and data for each row

for key, data in table.scan(row_prefix='row'):
    print key, data  # prints 'value1' and 'value2'

row = table.delete('row-key')

# For CentOS6 Users:
# Run these steps instead if you have only Python 2.6, because Python 2.7 at least is required,
# Otherwise, skip the entire section of comments and continue if you already have Python 2.7 as default

# yum install yum install -y centos-release-SCL
# yum install -y python27
# export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/rh/python27/root/usr/lib64/:/opt/rh/python27/root/usr/lib/
# /opt/rh/python27/root/usr/bin/easy_install pip
# /opt/rh/python27/root/usr/bin/pip install kerberos pure-sasl
# /opt/rh/python27/root/usr/bin/pip install git+https://github.com/QwertyManiac/happybase.git@0.9-security --upgrade
# /opt/rh/python27/root/usr/bin/python foo.py

~> easy_install pip
~> pip install pure-sasl kerberos
~> pip install git+https://github.com/QwertyManiac/happybase.git@0.9-security --upgrade
~> python foo.py`

**Comment by QwertyManiac** [comment]
Thriftpy request for adding TSaslClientTransport is open at https://github.com/eleme/thriftpy/issues/213

**Comment by wbolster** [comment]
sorry for not giving appropriate attention. is this still relevant with thriftpy?

**Comment by wbolster** [comment]
btw the older thrift lib is very unpythonic and requires code generation (yuck) using c++ tools spitting out boatloads of noise. also there have been various py3 issues (see older tickets about py3). hopefully that explains the switch.

**Comment by QwertyManiac** [comment]
Thanks for taking a look @wbolster, and on the comments why the migration was required.

Yes this remains a problem (even more so, in fact) with thriftpy (see eleme/thriftpy#213) as it does not carry a native SASL transport yet.

The patch works well applied on the 0.9 release though (pre-thriftpy), which I carry at the QwertyManiac/happybase.git's "0.9-security" fork branch.

**Comment by funes79** [comment]
Hi @QwertyManiac , I tried to test your modified version of happybase/thriftpy but it fails to import happybase:
```
 Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "build/bdist.linux-x86_64/egg/happybase/__init__.py", line 11, in <module>
  File "build/bdist.linux-x86_64/egg/happybase/connection.py", line 10, in <module>
ImportError: cannot import name TSaslClientTransport

```

**Comment by wbolster** [comment]
progress on this depends on
https://github.com/eleme/thriftpy/issues/213

once thriftpy supports it, happybase connections can grow a few appropriate flags to use it.

**Comment by zhiyajun11** [comment]
https://github.com/zhiyajun11/happybase-kerberos-patch 
support happybase>=1.0.0

File: happybase/connection.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 import logging
 
 from thriftpy.thrift import TClient
-from thriftpy.transport import TBufferedTransport, TFramedTransport, TSocket
+from thriftpy.transport import TBufferedTransport, TFramedTransport, TSocket, TSaslClientTransport
```
Comment (wbolster) [inline happybase/connection.py]: can you sort these imports?

:nerd_face: nitpick :)

File: happybase/connection.py [hunk 2]
```diff
@@ -80,6 +80,22 @@ class Connection(object):
...
+    only required if your HBase Thrift Server runs with a Kerberos principal
+    identity different than ``hbase``. Passing the hostname as part of the
+    `sasl_service_name` is not required, as it is inferred from the host passed
+    to the connection. Note that for using `use_kerberos` you need to also have
+    the Python modules ``puresasl`` and ``kerberos`` installed. The latter
+    module works only with Python 2.7.0+ due to reliance on Capsule features.
+    As such, this feature can only be used on Python 2.7 onwards.
+
+    .. versionadded:: 0.10
+       `use_kerberos` and `sasl_service_name` arguments
```
Comment (wbolster) [inline happybase/connection.py]: can you make this version `1.2.0` instead?

**Comment by wbolster** [comment]
hmmm i am actually confused now. does this cleanly work with `thriftpy`?

(i have almost no time for this project, sorry for the noise and confusion.)

**Comment by QwertyManiac** [comment]
No this patch was written for the version prior to thriftpy move. As such, I'm marking as closed as it is not applicable to the current future path of this project, and https://github.com/Thriftpy/thriftpy/issues/213 will enable this for happybase in future when it happens.

Currently I use a forked version prior to thriftpy move to get around this limitation, its working OK so far.

**Pull request closed without merging**